src/v/cluster/plugin_frontend.h

```diff
@@ -0,0 +1,166 @@
+    plugin_frontend(
+      model::node_id,
+      partition_leaders_table*,
+      plugin_table*,
+      topic_table*,
+      controller_stm*,
+      rpc::connection_cache*,
+      ss::abort_source*);
+
+    using notification_id = plugin_table::notification_id;
+    using notification_callback = plugin_table::notification_callback;
+    struct mutation_result {
+        // The UUID of the modified transform, mostly useful for cleanup
+        // of wasm_binaries.
+        // See transform_metadata::uuid
+        uuid_t uuid;
+        errc ec;
+    };
+
+    // Create or update a transform by name.
+    ss::future<errc> upsert_transform(
+      model::transform_metadata, model::timeout_clock::time_point);
+
+    // Remove a transform by name.
+    ss::future<mutation_result>
+      remove_transform(model::transform_name, model::timeout_clock::time_point);
+
+    // Register for updates going forward.
+    //
+    // This is a core local operation, you can register for updates to the
+    // transforms table. If you want to lookup the transform you can then do
+    // that in the callback.
+    notification_id register_for_updates(notification_callback);
+
+    // Unregister a previously registered notification.
+    void unregister_for_updates(notification_id);
+
+    // Lookup a transform by ID. Useful when paired with registering for
+    // updates.
+    std::optional<model::transform_metadata>
+      lookup_transform(model::transform_id) const;
+
+    // Lookup transforms for input topics.
+    absl::flat_hash_map<model::transform_id, model::transform_metadata>
+      lookup_transforms_by_input_topic(model::topic_namespace_view) const;
+
+    // Get a snapshot of all the transforms that currently exist.
+    absl::flat_hash_map<model::transform_id, model::transform_metadata>
+    all_transforms() const;
+
+private:
+    // Perform a mutation request, check if this node is the cluster leader and
+    // ensuring that the right core is used for local mutations.
+    ss::future<mutation_result>
+      do_mutation(transform_cmd, model::timeout_clock::time_point);
+
+    // If this node is not the leader, dispatch this mutation to `node_id` as
+    // the cluster's leader.
+    ss::future<mutation_result> dispatch_mutation_to_remote(
+      model::node_id, transform_cmd, model::timeout_clock::duration);
+
+    // Performs (and validates) a mutation command to be inserted into the
+    // controller log.
+    //
+    // This must take place on the controller shard on the cluster leader to
+    // ensure consistency.
+    ss::future<mutation_result> do_local_mutation(
+      transform_cmd, model::offset, model::timeout_clock::time_point);
+
+    // If this command is a upsert, then the ID is assigned.
+    //
+    // The offset supplied is the latest offset applied in raft0
+    //
+    // This must take place on the controller shard on the cluster leader to
+    // ensure consistency.
+    void assign_id(transform_cmd*, model::offset);
```

Comment:
elsewhere in controller we avoid this id assignment by making id = offset of the transform_cmd in the controller log (after replication is successful).  That way we don't need to assign them upfront, which is also error prone I think.

Reply:
See my note in the overall feedback, but I think this means the ID changes as there are updates, which I would like stable IDs.

Reply:
discussed offline, I think we can assign ids _after_ replication in the stm as it applies depending on if a transform already exists.

Reply:
Done, this is much better thanks for helping me understand the issues here :pray: